### ISSUE TITLE
Suggested New Approach to Context in RiGrammar

### ISSUE THREAD

**Issue by srt19170** [dhowe/RiTa#502]
RiGrammar executes callbacks in grammar rules by using eval() first in a global context, and then in a context provided by the user (as a second argument to expandFrom).  This isn't well documented, but this context is expected to be either a single function or an object with each key a function.  If the global eval of the callback fails, these functions are then added to the scope of the eval() and the eval is tried again.

This approach has a number of limitations.  Most importantly (1) you can only provide context functions, and not (say) a binding of an object to a name, and (2) the context functions themselves are evaluated in the global context.  These problems can generally be worked around, but make working within the local scope of the code calling the grammar difficult.

A better approach would be to provide an optional closure as context for the evaluation:

`return riGrammar.expandFrom('<start>', function (str) { return eval(str); });`

which can then be used directly to eval the callback function in the calling functions closure

`res = context(exec);`

Semantically this is much easier for the user -- the callback executes exactly as it would if it were executed at the point the grammar is expanded.

**Comment by dhowe** [comment]
Thanks for this So you're suggesting trying the global context first, then the optional closure, correct? At quick glance this seems to make good sense. Any thoughts on how to possibly include a similar mechanism for java (perhaps via lambdas)? Also, if you wanted to submit a PR with some tests, that would be great

**Comment by srt19170** [comment]
I think if context is provided the global execution should be skipped.  If the user has supplied a context it's clear that he wants to execute callbacks in that context.  And there's a small chance that that user is providing a context in order to shadow some functions in the global context -- so if you execute the global context first, it might succeed but return the "wrong" answer.

I'll try to submit a PR and tests.  I've actually implemented this already in my own copy of RiTa -- it's not a huge change.  (And I think with this approach the entire Scope object goes away, although I haven't checked to see if it is used elsewhere as I'm only using RiGrammar.)

**Comment by srt19170** [comment]
Another enhancement that I find useful is to allow a callback in the weighting of a rule, e.g., 

     '<city>': '<city1> [x1] | <city2> [4]'

In this case, when the rule engine encounters a rule weight that isNaN(), it tries to evaluate the weight as an expression in the provided (falling back to global) context.  This allows dynamic weighting of rule options, and is also handy for implementing conditional rules, e.g., 

     '<city>': '<river-city> [isaRiver()] | <default-city>'

Here the function isaRiver() can return 0 when the city is not on a river, so that option will be "turned off".

Implementing this is pretty simple:  PROB_PATT has to be modified to allow non-numeric weights, and getStochasticRule has to be modified to attempt to eval temp[name] if that isNaN().  If you allow the user to pass in a local context (as above) then you also need to pass that down into getStochasticRule.  Note that in this case no backticks are required to distinguish the callback, but they could be kept for consistency if you think that's worthwhile.

**Comment by dhowe** [comment]
I like these suggestions. The complicating factor has been the shared API between Javascript and Java.; the idea being that whenever possible functions would take the same arguments types in both. As best I can see there's no nice way to allow this in Java (functional interface perhaps?). But it might be worth considering it as a special case regardless, especially as the current behavior doesn't work correctly in Node (https://github.com/dhowe/RiTa/issues/71).

**Comment by srt19170** [comment]
Okay, I've create https://github.com/dhowe/RiTaJS/pull/98 to address the first part of this issue.  (This is the first time I've created a PR, so apologies if I've done anything wrong.)

The PR adds closures for callbacks as described above, except that it tries the local context (closure) first if it is provided.  This permits the user to shadow a global function with a local context.  (Probably an edge case, but the reverse might lead to some hard-to-debug errors.)

I've provided two additional test cases that test failure and success of local context.  Note that all of the existing test cases continue to work because they work in the global context.

Once this is accepted I will provide a PR for a similar change in rule weighting.
